tests/test_SimCore.cpp

```diff
@@ -0,0 +1,242 @@
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#define CATCH_CONFIG_MAIN
+
+// System headers
+#include <exception>
+#include <thread>
+
+// 3rd party headers
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/catch_session.hpp>
+
+// Project headers
+#include "control/PowerSubsystem.h"
+#include "simulator/SimCore.h"
+#include "system/Config.h"
+#include "system/Globals.h"
+#include "util/Log.h"
+
+using namespace std;
+using namespace LSST::m2cellcpp::simulator;
+using namespace LSST::m2cellcpp::control;
+using namespace LSST::m2cellcpp;
+
+void testPowerOn(PowerSubsystemConfig const& psCfg, double voltage, double current,
+        string const& note, bool& success, bool& done) {
+    if (voltage > psCfg.getMinVoltageFault()) {
+        if (voltage < psCfg.getMaxVoltageFault()) {
+            if (current > psCfg.getMaxCurrentFault()) {
+                LERROR(note, "Current=", current, " too high");
+                success = false;
+                done = true;
+            }
+            success = true;
+            done = true;
+        } else {
+            LDEBUG(note, "Voltage=", voltage, " too low");
+        }
+    } else {
+        LDEBUG(note, "Voltage=", voltage, " too high");
+    }
+}
+
+void testPowerOff(PowerSubsystemConfig const& psCfg, double voltage,
+        string const& note, bool& success, bool& done) {
+    if (voltage <= psCfg.getVoltageOffLevel()) {
+        success = true;
+        done = true;
+    } else {
+        LDEBUG(note, "Voltage=", voltage, " above offLevel");
+    }
+}
+
+void checkTimeout(util::CLOCK::time_point tStart, double maxTime, bool& done, string const& note) {
+    // check timeout
+    util::CLOCK::time_point tEnd = util::CLOCK::now();
+    const double timeDiff = util::timePassedSec(tStart, tEnd);
+    if (timeDiff > maxTime) {
+        LERROR(note, "Voltage timeout timeDiff=", timeDiff, " max=", maxTime);
+        done = true;
+    }
+}
+
+bool motorPowerOnTest(SimCore& simCore, PowerSubsystemConfig const& motorCfg) {
+    util::CLOCK::time_point tStart = util::CLOCK::now();
+    bool success = false;
+    bool done = false;
+    while (!done) {
+        auto info = simCore.getSimInfo();
+        if (info.outputPort.getBit(OutputPortBits::ILC_MOTOR_POWER_ON) == false) {
+            LDEBUG("motor power not on");
+        }
+        testPowerOn(motorCfg, info.motorVoltage, info.motorCurrent, "motor", success, done);
+        checkTimeout(tStart, motorCfg.outputOnMaxDelay(), done, "motor");
+        this_thread::sleep_for(0.04s);
+    }
```

Comment:
Is it possible to have the edge case that the `done` is always `false` to run the loop forever?

Reply:
I don't think that's an issue `tStart` is set to `now()`  and the timeouts are only a few seconds, so `checkTimeout` should always set `done = true` in a reasonable amount of time.

If the timeout is unreasonable, then this is a unit test and the place where it should cause a problem.